src/grammar.ts

```diff
@@ -1436,6 +1436,24 @@ export class StackElement implements StackElementDef {
 	public hasSameRuleAs(other: StackElement): boolean {
 		return this.ruleId === other.ruleId;
 	}
+
+	public hasEndlessLoop(anchorPosition: number): boolean {
+		if (anchorPosition === -1) {
+			// method is unavailable
+			return true;
+		}
+
+		let se = this as StackElement;
+
+		while (anchorPosition === se._anchorPos && se.parent !== null) {

```

Comment:
I'm a bit concerned about adding this `while` loop. The two checks were effectively `O(1)` before and would happen only if the stream did not advance. Now this could potentially walk a long time through the stack linked list, and it would do so twice per iteration.

Reply:
It certainly should be thought about, maybe even performance tested to validate what affects it has on some complicated grammars (those with zero length match based begin/end rules) with large files.

However, keep in mind that the while loop only iterates if and while the anchor hasn't advanced,

Doing a quick comparison between this PR and the master branch, using `measure-command` against the `inspect` script with the grammar I am working on with a 650 line source code file, the script time was 2.15 seconds average either way.  I also measured both ways with the 'grammar.ts' file and the typescript grammar, and I could not discern a difference in the noise (both varied from 4.65 to 5 seconds each).  The biggest impact would occur on grammars that utilize multiple levels of zero length match Begin/End rules for common content.

Worst case, the only test needed to eliminate the base issue is to compare the current anchor position to the stacked anchor position (instead of using `hasAdvanced`) before testing the `ruleId` values of the last two stacked rules.

Reply:
Maybe I should be using `_enterPos`?  Maybe the first test isn't needed, or should return false?

At least this still works the same:

```typescript
	public hasEndlessLoop(): boolean {
		if (this._enterPos === -1) {
			// method is unavailable
			return true;
		}

		let se = this.parent;

		while (se !== null && this._enterPos == se._enterPos) {
			if (this.hasSameRuleAs(se)) {
				// endless loop detected
				return true;
			}
			se = se.parent;
		}
		return false;
	}
```

Reply:
:+1: This makes sense to me.